crates/wasi-nn/src/tf.rs

```diff
@@ -0,0 +1,295 @@
+//! Implements the wasi-nn API.
+use crate::api::{Backend, BackendError, BackendExecutionContext, BackendGraph};
+use crate::witx::types::{ExecutionTarget, GraphBuilderArray, Tensor, TensorType};
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::path::Path;
+use std::str;
+use std::sync::Arc;
+use tensorflow::{
+    FetchToken, Graph, Operation, SavedModelBundle, SessionOptions, SessionRunArgs, Status,
+};
+use tensorflow::{SignatureDef, Tensor as TFTensor};
+use tensorflow::{TensorInfo, DEFAULT_SERVING_SIGNATURE_DEF_KEY};
+
+#[derive(Default)]
+pub(crate) struct TensorflowBackend();
+
+impl Backend for TensorflowBackend {
+    fn name(&self) -> &str {
+        "tensorflow"
+    }
+
+    fn load(
+        &mut self,
+        builders: &GraphBuilderArray<'_>,
+        _target: ExecutionTarget,
+        map_dir: &Option<Vec<(String, String)>>,
+    ) -> Result<Box<dyn BackendGraph>, BackendError> {
+        if let Some(dir) = map_dir {
+            if builders.len() != 2 {
+                return Err(BackendError::InvalidNumberOfBuilders(2, builders.len()).into());
+            }
+            // Initialize the Tensorflow backend
+            let _retval = tensorflow::library::load().or_else(|e| {
+                println!("Error loading the TensorFlow backend: \n {}", e);
+                Err(e)
+            });
+
+            let mut graph = Graph::new();
+            let builders = builders.as_ptr();
+            let guest_map = builders.read()?.as_slice()?;
+            let guest_map_str = str::from_utf8(&guest_map).unwrap();
+            let exp_dir = builders.add(1)?.read()?.as_slice()?;
+            let exp_str = str::from_utf8(&exp_dir).unwrap();
+
+            // Don't allow navigation outside of the sandbox
+            if !exp_str.contains("..") {
+                for i in 0..dir.len() {
+                    if dir[i].0 == guest_map_str {
+                        //Append the stored mapdir path with the user path.
+                        let full_path = std::fs::canonicalize(
+                            Path::new(&dir[i].1.clone()).join(Path::new(exp_str)),
+                        );
+
+                        //Check that path actually exists
+                        let full_path = match full_path {
+                            Ok(fp) => fp,
+                            Err(_e) => return Err(BackendError::MissingMapDir()),
+                        };
+
+                        let bundle = SavedModelBundle::load(
+                            &SessionOptions::new(),
+                            &["serve"],
```

Comment:
This seems to be information injected when the model was saved; I don't think it will always be `"serve"`?